core/src/main/java/org/pentaho/di/core/DBCache.java

```diff
@@ -0,0 +1,231 @@
+    try {
+      clear( null );
+
+      // Serialization support for the DB cache
+      //
+      log = new LogChannel( "DBCache" );
+
+      String filename = fileNameSupplier.get();
+      File file = new File( filename );
+      if ( file.canRead() ) {
+        log.logDetailed( "Loading database cache from file: [" + filename + "]" );
+
+        try ( DataInputStream dis = new DataInputStream( new FileInputStream( file ) ) ) {
+          loadFileToCache( dis );
+        }
+      } else {
+        log.logDetailed( "The database cache doesn't exist yet." );
+      }
+    } catch ( Exception e ) {
+      throw new KettleFileException( "Couldn't read the database cache", e );
+    }
+  }
+
+  @SuppressWarnings( { "squid:S2189", "squid:S1451" } )
+  private void loadFileToCache( DataInputStream dis ) throws KettleFileException, SocketTimeoutException {
+    int counter = 0;
+    try {
+      //noinspection InfiniteLoopStatement Only way to detect EOF on DataInputStream is with exception
+      while ( true ) {
+        DBCacheEntry entry = new DBCacheEntry( dis );
+        RowMetaInterface row = new RowMeta( dis );
+        cache.put( entry, row );
+        counter++;
+      }
+    } catch ( KettleEOFException eof ) {
+      log.logDetailed( "We read " + counter + " cached rows from the database cache!" );
+    }
+  }
+
+  public void saveCache() throws KettleFileException {
+    try {
+      // Serialization support for the DB cache
+      //
+      String filename = fileNameSupplier.get();
+      File file = new File( filename );
+      if ( !file.exists() || file.canWrite() ) {
+
+        try ( DataOutputStream dos =
+                new DataOutputStream( new BufferedOutputStream( new FileOutputStream( file ), 10000 ) ) ) {
+
+          int counter = 0;
+
+          for ( DBCacheEntry entry : cache.keySet() ) {
+            entry.write( dos );
+
+            // Save the corresponding row as well.
+            RowMetaInterface rowMeta = get( entry );
+            if ( rowMeta != null ) {
+              rowMeta.writeMeta( dos );
+              counter++;
+            } else {
+              throw new KettleFileException( "The database cache contains an empty row. We can't save this!" );
+            }
+          }
+          log.logDetailed( "We wrote " + counter + " cached rows to the database cache!" );
+        }
+      } else {
+        throw new KettleFileException( "We can't write to the cache file: " + filename );
+      }
+    } catch ( Exception e ) {
+      throw new KettleFileException( "Couldn't write to the database cache", e );
+    }
+  }
+
+  /**
+   * Create the database cache instance by loading it from disk
+   *
+   * @return the database cache instance.
+   */
+  @SuppressWarnings( "squid:S00112" )
+  public static DBCache getInstance() {
+    if ( dbCache != null ) {
+      return dbCache;
+    }
+    try {
+      dbCache = new DBCache();
+    } catch ( KettleFileException kfe ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)